src/synthesizer/typescript_synthesizer.rs

```diff
@@ -229,13 +267,17 @@ pub fn to_string_ir(resource_value: &ResourceIr) -> Option<String> {
                 None => String::from("{}"),
                 Some(x) => x,
             };
-
             let false_expr = match to_string_ir(false_expr) {
                 None => String::from("{}"),
                 Some(x) => x,
             };
 
-            Option::Some(format!("({})?{}:{}", bool_expr, true_expr, false_expr))
+            if false_expr == "{}" || false_expr == "[]" {
+                // false_expr here ("{}" or "[]") might give a type mismatch error so we remove it.
+                Option::Some(format!("{}", true_expr))
+            } else {
+                Option::Some(format!("{} ? {} : {}", bool_expr, true_expr, false_expr))
+            }
         }
```

Comment:
The problem is that actually changes semantics of the if statement -- and we can't compress to only the true value

As am example:

```
{ "Fn::If" ["OnDemand", { "ProvisionedWriteThroughput": 10 }, "AWS::NoValue"] }
 ```

No Value turns into a false_expr of `{}` -- which is technically incorrect as well -- as I learned through experimentation 😂. but translating that to only the true expression would actually be an incorrect program and lead to an outage in certain scenarios (imagine instead of swapping to null above, it just always output ProvisionedWriteThroughput).

In this case, I think we should swap false_expr to return `undefined`. Props would already have a union of undefined in cases where a user can optionally supply values. This preserves the original semantics, and will lead to correct TS programs -- opposed to `{}` which is an `Any` type -- which would not satisfy interfaces.

Reply:
Also, for the [] -- I think in all cases an empty array would satisfy any type that required an array so we can just keep that one.

Reply:
I honestly think that means the if statement is unreachable (as in if the if statement went on the false path, the cfn template would fail). Sadly, noctilucent can't detect that as it's a semantic error.